### ISSUE TITLE
System timezone offset not picked up?

### ISSUE THREAD

**Issue by arbrebinaire** [jtmoon79/super-speedy-syslog-searcher#83]
Hi,

Fantastic tool: hard to find but absolutely irreplaceable.  Thank you for your work.

My system just changed its time offset from EST to EDT.  Accordingly, the Apache logs now use `-0400` as expected.

This command:

`s4 -a 20230315T105500 -b 20230315T110500 /var/log/virtualmin/*_access_log`

produces this final log line (+ some before 12:05):

`64.124.8.21 - - [15/Mar/2023:12:04:54 -0400] "GET /en/shop/lynx-en/525/ HTTP/1.1" 200 72553 "-" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/61.0.3163.79 Safari/537.36"`

Seems to me s4 is not picking up the local system time offset:

```
[sysadmin@acer ~]# sudo timedatectl
Local time: Thu 2023-03-16 09:24:02 EDT
Universal time: Thu 2023-03-16 13:24:02 UTC
RTC time: Thu 2023-03-16 13:24:01
Time zone: America/New_York (EDT, -0400)
System clock synchronized: yes
NTP service: active
RTC in local TZ: no
```

The command:

`s4 -a 20230315T105500 -b 20230315T110500 -t EDT /var/log/virtualmin/*_access_log`

yields this final log line  (+ some before 11:05):

`204.193.36.5 - - [15/Mar/2023:11:04:52 -0400] "GET /catalog15/b.pdf HTTP/1.1" 304 -  "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/110.0.0.0 Safari/537.36"`

The time window is correct when specifying EDT: not the case without the `-t` parameter set to the correct time zone.

What am I doing wrong?

Thanks much.

**Comment by jtmoon79** [comment]
***tl;dr*** adjust your datetime filters.

<br/>

Assuming you're using the latest version, 0.3.53.

### Context

IIUC, the log file looks like

```lang-text
204.193.36.5 - - [15/Mar/2023:11:04:52 -0400] "GET /catalog15/b.pdf HTTP/1.1" 304 -  "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/110.0.0.0 Safari/537.36"
64.124.8.21 - - [15/Mar/2023:12:04:54 -0400] "GET /en/shop/lynx-en/525/ HTTP/1.1" 200 72553 "-" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/61.0.3163.79 Safari/537.36"
```

You're suggesting a Daylight Savings Time change occurred between those two log lines. However, the timezone offset should be `-0500` in the first log message if a DST change occurred (i.e. if the system was previously in the EST -0500 timezone offset).

***Are you sure you copy+pasted the expected log lines?***

### Explanation -t

Passing `-t` will not make a difference here. The option `-t` is for setting the timezone offset for _unknown_ timezone offsets. e.g. if those datetime stamps did not have substring `-0400`.
But the datetime stamps do include timezone offsets so `-t` has no effect.

### Check -u

You can double-check how `s4` interprets the log message datetime stamp by passing `-u` or `-l`.

In my test I passed `-u` to cause preprint of each datetime stamp as UTC
```lang-text
‣ s4 -u /tmp/83.log
20230315T150452.000+0000:204.193.36.5 - - [15/Mar/2023:11:04:52 -0400] "GET /catalog15/b.pdf HTTP/1.1" 304 -  "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/110.0.0.0 Safari/537.36"
20230315T160454.000+0000:64.124.8.21 - - [15/Mar/2023:12:04:54 -0400] "GET /en/shop/lynx-en/525/ HTTP/1.1" 200 72553 "-" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/61.0.3163.79 Safari/537.36"
```

### Datetime filters

> Seems to me s4 is not picking up the local system time offset

Try passing datetime filters with the timezone you want. So _I think you want_ timezone EDT (`-0400`) and _I think you want_ to adjust the hour by one.

So run:
```lang-text
‣ s4 -a 20230315T115500-0400 -b 20230315T120500-0400 /tmp/83.log
``` 

For me this printed:

```lang-text
64.124.8.21 - - [15/Mar/2023:12:04:54 -0400] "GET /en/shop/lynx-en/525/ HTTP/1.1" 200 72553 "-" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/61.0.3163.79 Safari/537.36"
```

If I'm misunderstanding, go ahead and zip the log file and post somewhere online.

**Comment by arbrebinaire** [comment]
A few details to respond to your comment:

- I was using version 0.2.47; just upgraded to 0.3.53;
-  I'm afraid s4@0.3.53 behaves the very same way (with an automatic offset... offset by 1 hour exactly when compared with the server time);
- The EST -> EDT time change occurred on the 12th of March at 02:00:00 (which became 03:00:00). s4 (with stated time window parameters as you can read in my original post) was run on a single log file with every line having a time offset of -04:00. I reran s4@0.3.53 on a single log file to ensure consistency (the original s4 call was made on multiple log files), with the same result as mentioned above (please see the link below to access that single file);
- I thought the -t option had the effect of modifying the default offset option which is set to the system UTC offset. According to the man page the `Default value is the local system timezone offset.`  (that's why I ran the `timedatectl` utility as shown above to make sure the system offset was correct).  I came to the conclusion that maybe setting the -t option to EDT would filter the proper lines: it does with s4@0.3.53 as with s4@0.2.47.  I guess I'm confused by the man page -t option entry.  I'm confused furthermore by your explanation since the -t EDT does change s4 behavior, with proper results.  The fact that the default offset of s4, according to the man page, is the system offset is useful: time changes would not affect any and all scripted s4 call that do not mention the offset.

I include the original single log file on which I tested s4 (very few lines were filtered out for security reasons):
[domain.com.log_access_log.txt](https://github.com/jtmoon79/super-speedy-syslog-searcher/files/10997201/domain.com.log_access_log.txt)

I hope I clarified my original post.  If not, please ask me any info you need.

Thanks again.

**Comment by jtmoon79** [comment]
Thanks for providing the example files. That helps.

> time changes would not affect any and all scripted s4 call that do not mention the offset.

Yes, correct.

<br/>

### Expected vs. Actual

I'm afraid I don't understand what the problem is.

Can you provide the following:

1. Command run using the linked file `domain.com.log_access_log.txt` (in a code block)
2. Expected output (in a separate fenced code block, e.g. \`\`\` )
3. Actual output (in a separate fenced code block, e.g. \`\`\` )

So for example,

1. `s4 -a 20230315T115500 -b 20230315T120500 domain.com.log_access_log`
2. expect
```lang-text
107.178.195.12 - - [15/Mar/2023:15:55:47 -0400] "GET /57bu5e4a36 HTTP/1.1" 200 15089 "-" "Mozilla/5.0 (compatible; Google-Apps-Script; beanserver; +https://script.google.com; id: UAEmdDd8omn1_3KV_TP-n0k9Wf_q4uUv1-g)"
107.178.194.226 - - [15/Mar/2023:15:56:47 -0400] "GET /57bu5e4a36 HTTP/1.1" 200 15089 "-" "Mozilla/5.0 (compatible; Google-Apps-Script; beanserver; +https://script.google.com; id: UAEmdDd8omn1_3KV_TP-n0k9Wf_q4uUv1-g)"
```
3. actual
```lang-text
107.178.195.12 - - [15/Mar/2023:14:55:49 -0400] "GET /57bu5e4a36 HTTP/1.1" 200 15089 "-" "Mozilla/5.0 (compatible; Google-Apps-Script; beanserver; +https://script.google.com; id: UAEmdDd8omn1_3KV_TP-n0k9Wf_q4uUv1-g)"
35.187.143.20 - - [15/Mar/2023:14:57:48 -0400] "GET /57bu5e4a36 HTTP/1.1" 200 15089 "-" "Mozilla/5.0 (compatible; Google-Apps-Script; beanserver; +https://script.google.com; id: UAEmdDd8omn1_3KV_TP-n0k9Wf_q4uUv1-g)"
```

### s4 default offset

Please copy+paste the output of this shell command (into a fenced code block, e.g. \`\`\` ):
```
s4 --help 2>&1 | grep -A 6 -Fe '--tz-offset'
```

**Comment by arbrebinaire** [comment]
Here we go:

### Expected vs. Actual

Command:
`s4 -a 20230315T115500 -b 20230315T120500 /home/sysadmin/domain.com.log_access_log`

Expected:
```107.178.195.63 - - [15/Mar/2023:11:55:48 -0400] "GET /57bu5e4a36 HTTP/1.1" 200 15089 "-" "Mozilla/5.0 (compatible; Google-Apps-Script; beanserver; +https://script.google.com; id: UAEmdDd8omn1_3KV_TP-n0k9Wf_q4uUv1-g)"
35.187.143.20 - - [15/Mar/2023:11:56:48 -0400] "GET /57bu5e4a36 HTTP/1.1" 200 15089 "-" "Mozilla/5.0 (compatible; Google-Apps-Script; beanserver; +https://script.google.com; id: UAEmdDd8omn1_3KV_TP-n0k9Wf_q4uUv1-g)"
34.116.22.59 - - [15/Mar/2023:11:57:48 -0400] "GET /57bu5e4a36 HTTP/1.1" 200 15089 "-" "Mozilla/5.0 (compatible; Google-Apps-Script; beanserver; +https://script.google.com; id: UAEmdDd8omn1_3KV_TP-n0k9Wf_q4uUv1-g)"
107.178.193.226 - - [15/Mar/2023:11:58:53 -0400] "GET /57bu5e4a36 HTTP/1.1" 200 15089 "-" "Mozilla/5.0 (compatible; Google-Apps-Script; beanserver; +https://script.google.com; id: UAEmdDd8omn1_3KV_TP-n0k9Wf_q4uUv1-g)"
34.116.22.55 - - [15/Mar/2023:11:59:48 -0400] "GET /57bu5e4a36 HTTP/1.1" 200 15089 "-" "Mozilla/5.0 (compatible; Google-Apps-Script; beanserver; +https://script.google.com; id: UAEmdDd8omn1_3KV_TP-n0k9Wf_q4uUv1-g)"
107.178.195.63 - - [15/Mar/2023:12:00:48 -0400] "GET /57bu5e4a36 HTTP/1.1" 200 15089 "-" "Mozilla/5.0 (compatible; Google-Apps-Script; beanserver; +https://script.google.com; id: UAEmdDd8omn1_3KV_TP-n0k9Wf_q4uUv1-g)"
34.116.22.55 - - [15/Mar/2023:12:01:48 -0400] "GET /57bu5e4a36 HTTP/1.1" 200 15089 "-" "Mozilla/5.0 (compatible; Google-Apps-Script; beanserver; +https://script.google.com; id: UAEmdDd8omn1_3KV_TP-n0k9Wf_q4uUv1-g)"
34.116.22.59 - - [15/Mar/2023:12:02:48 -0400] "GET /57bu5e4a36 HTTP/1.1" 200 15089 "-" "Mozilla/5.0 (compatible; Google-Apps-Script; beanserver; +https://script.google.com; id: UAEmdDd8omn1_3KV_TP-n0k9Wf_q4uUv1-g)"
107.178.195.34 - - [15/Mar/2023:12:03:48 -0400] "GET /57bu5e4a36 HTTP/1.1" 200 15089 "-" "Mozilla/5.0 (compatible; Google-Apps-Script; beanserver; +https://script.google.com; id: UAEmdDd8omn1_3KV_TP-n0k9Wf_q4uUv1-g)"
34.116.22.57 - - [15/Mar/2023:12:04:51 -0400] "GET /57bu5e4a36 HTTP/1.1" 200 15089 "-" "Mozilla/5.0 (compatible; Google-Apps-Script; beanserver; +https://script.google.com; id: UAEmdDd8omn1_3KV_TP-n0k9Wf_q4uUv1-g)"
```

Actual:
```107.178.195.12 - - [15/Mar/2023:12:55:48 -0400] "GET /57bu5e4a36 HTTP/1.1" 200 15089 "-" "Mozilla/5.0 (compatible; Google-Apps-Script; beanserver; +https://script.google.com; id: UAEmdDd8omn1_3KV_TP-n0k9Wf_q4uUv1-g)"
107.178.195.10 - - [15/Mar/2023:12:56:48 -0400] "GET /57bu5e4a36 HTTP/1.1" 200 15089 "-" "Mozilla/5.0 (compatible; Google-Apps-Script; beanserver; +https://script.google.com; id: UAEmdDd8omn1_3KV_TP-n0k9Wf_q4uUv1-g)"
35.243.21.29 - - [15/Mar/2023:12:57:54 -0400] "GET /57bu5e4a36 HTTP/1.1" 200 15089 "-" "Mozilla/5.0 (compatible; Google-Apps-Script; beanserver; +https://script.google.com; id: UAEmdDd8omn1_3KV_TP-n0k9Wf_q4uUv1-g)"
34.116.22.59 - - [15/Mar/2023:12:58:48 -0400] "GET /57bu5e4a36 HTTP/1.1" 200 15089 "-" "Mozilla/5.0 (compatible; Google-Apps-Script; beanserver; +https://script.google.com; id: UAEmdDd8omn1_3KV_TP-n0k9Wf_q4uUv1-g)"
107.178.193.226 - - [15/Mar/2023:12:59:47 -0400] "GET /57bu5e4a36 HTTP/1.1" 200 15089 "-" "Mozilla/5.0 (compatible; Google-Apps-Script; beanserver; +https://script.google.com; id: UAEmdDd8omn1_3KV_TP-n0k9Wf_q4uUv1-g)"
107.178.194.248 - - [15/Mar/2023:13:00:48 -0400] "GET /57bu5e4a36 HTTP/1.1" 200 15089 "-" "Mozilla/5.0 (compatible; Google-Apps-Script; beanserver; +https://script.google.com; id: UAEmdDd8omn1_3KV_TP-n0k9Wf_q4uUv1-g)"
107.178.193.226 - - [15/Mar/2023:13:01:48 -0400] "GET /57bu5e4a36 HTTP/1.1" 200 15089 "-" "Mozilla/5.0 (compatible; Google-Apps-Script; beanserver; +https://script.google.com; id: UAEmdDd8omn1_3KV_TP-n0k9Wf_q4uUv1-g)"
35.243.21.29 - - [15/Mar/2023:13:02:47 -0400] "GET /57bu5e4a36 HTTP/1.1" 200 15089 "-" "Mozilla/5.0 (compatible; Google-Apps-Script; beanserver; +https://script.google.com; id: UAEmdDd8omn1_3KV_TP-n0k9Wf_q4uUv1-g)"
34.116.22.59 - - [15/Mar/2023:13:03:48 -0400] "GET /57bu5e4a36 HTTP/1.1" 200 15089 "-" "Mozilla/5.0 (compatible; Google-Apps-Script; beanserver; +https://script.google.com; id: UAEmdDd8omn1_3KV_TP-n0k9Wf_q4uUv1-g)"
107.178.194.252 - - [15/Mar/2023:13:04:48 -0400] "GET /57bu5e4a36 HTTP/1.1" 200 15089 "-" "Mozilla/5.0 (compatible; Google-Apps-Script; beanserver; +https://script.google.com; id: UAEmdDd8omn1_3KV_TP-n0k9Wf_q4uUv1-g)"
```

What is expected in other words: the output matches the time window specified by the -a, -b options in reference to the system offset.  If the system offset is -04:00 and the time period is from T11:55:00 to T12:05:00 print all log lines created between 11:55:00 and 12:05:00.

At the moment, s4 prints all log lines between 12:55:00 to 13:05:00 even though both the system time offset and the log lines have the exact same time offset of -04:00.  s4 prints log lines in the future :-)

### s4 default offset

Command:
`s4 --help 2>&1 | grep -A 6 -Fe '--tz-offset'`

```
  -t, --tz-offset <TZ_OFFSET>
          Default timezone offset for datetimes without a timezone.
          For example, "-0800", "+02:00", or "EDT". Ambiguous named timezones
          parsed from logs will use this value, e.g. timezone "IST" (to pass a
          value with leading "-", quote the argument, e.g. "-t=-0800").
          Default value is the local system timezone offset. [default: -05:00]
  -u, --prepend-utc
```

For completeness:

Output of the date command on the same system
`date +%Z`
```
EDT
```
`date +%z`
```
-0400
```

For all I can tell s4 believes the system time offset is still -05:00 when it is actually -04:00 as shown by both date commands above.

Thank you for reading through my broken English!

**Comment by arbrebinaire** [comment]
A bit off topic:

I have no idea what the hurdles are but this tool is so advanced, so powerful and so useful that I would suggest offering it to the GNU Project.

In my mind this is like rsync: a seminal "Swiss Army knife" to be used every day by hundreds of thousands sysadmins.  It should have a broader audience.

All the best,

François
